src/main/java/org/zalando/nakadi/service/timeline/TimelineSyncImpl.java

```diff
@@ -0,0 +1,365 @@
+                        }
+                    }
+                }
+            }
+            try {
+                zooKeeperHolder.get().setData()
+                        .forPath(toZkPath("/nodes/" + thisId), String.valueOf(change.version).getBytes(CHARSET));
+            } catch (final Exception ex) {
+                log.error("Failed to update node version in zk. Will try to reprocess again", ex);
+                return;
+            }
+            queuedChanges.poll();
+            log.info("Delayed change {} successfully processed", change);
+        }
+    }
+
+    /**
+     * MUST be executed under zk lock. Called when version is changed
+     */
+    private void refreshVersion() {
+        try {
+            log.info("Adding refresh call to delayed changes list");
+            queuedChanges.add(
+                    new DelayedChange(readData("/version", this::versionChanged, Integer::parseInt),
+                            this.zooKeeperHolder.get().getChildren().forPath(toZkPath("/locked_et"))));
+        } catch (final RuntimeException ex) {
+            throw ex;
+        } catch (final Exception ex) {
+            throw new RuntimeException(ex);
+        }
+    }
+
+    private void versionChanged(final WatchedEvent ignore) {
+        refreshVersion();
+    }
+
+    private void runLocked(final Runnable action) {
+        try {
+            Exception releaseException = null;
+            final InterProcessLock lock =
+                    new InterProcessSemaphoreMutex(zooKeeperHolder.get(), ROOT_PATH + "/lock");
+            lock.acquire();
+            try {
+                action.run();
+            } finally {
+                try {
+                    lock.release();
+                } catch (final Exception ex) {
+                    log.warn("Failed to release lock", ex);
+                    releaseException = ex;
+                }
+            }
+            if (null != releaseException) {
+                throw releaseException;
+            }
+        } catch (final RuntimeException ex) {
+            throw ex;
+        } catch (final Exception ex) {
+            throw new RuntimeException(ex);
+        }
+    }
+
+    @Override
+    public Closeable workWithEventType(final String eventType) throws InterruptedException {
+        synchronized (localLock) {
+            while (lockedEventTypes.contains(eventType)) {
+                localLock.wait();
+            }
+            if (!eventsBeingPublished.containsKey(eventType)) {
+                eventsBeingPublished.put(eventType, 1);
+            } else {
+                eventsBeingPublished.put(eventType, 1 + eventsBeingPublished.get(eventType));
+            }
+        }
+        return () -> {
+            synchronized (localLock) {
+                final int currentCount = eventsBeingPublished.get(eventType);
+                if (1 == currentCount) {
+                    eventsBeingPublished.remove(eventType);
+                    localLock.notifyAll();
```

Comment:
@v-stepanov @lmontrieux 
There are two collections, access to which is synchronized using 1 lock (`localLock`): `eventsBeingPublished` and `lockedEventTypes`. When any of this collections is modified i notify about that using this lock.
Lock is not used for particular event type. Instead it is used for collection of event types.
@v-stepanov It is possible to have several event type timeline updates for different event types and for the same event type at the same moment.

Probably we should have offline review session for this.